# 👻  **Taller 9 - Machine Learning - SVM**
 <hr>
<img src="https://www.udistrital.edu.co/themes/custom/versh/images/default/preloader.png" align="left" width="192px" height="192px"/>
<img align="left" width="0" height="192px" hspace="10"/>
 
> Juan Felipe Rodríguez Galindo  - **COD. 20181020158**
<br></br>
[![Juan Felipe Rodriguez Galindo](https://img.shields.io/badge/Juferoga-github-br?style=flat-square)](https://gitlab.com/Juferoga)
[![License](https://img.shields.io/badge/License-GPL_V.3-blue?style=flat-square)](https://www.gnu.org/licenses/gpl-3.0.html)
 
 
Introducción a la Ciencia de Datos
 
Desarrrollo del taller propuesto en la clase de introducción a la ciencia de datos.

# ☠️ Crear función para variar kernels ☠️

<hr>

## Taller 
 
```
MACHINE LEARNING - SVM
 
1. Seleccionar 3 datasets de características diferentes

2. Para cada uno de los dataset evaluar un clasificador SVM que
   utilice diferentes kernels (RBF, Polinomial, Lineal, Sigmoidal
   con diferentes hiper-parámetros de configuración del kernel.

3. Sacar conclusiones. 
```

<hr>

## **Desarrollo**

1. Seleccionamos para el primer punto del taller los conjuntos de datos:

| *ID* 	|                        **Nombre**                       	| **Instancias** 	| **Atributos** 	|                                              **URL**                                              	|
|------	|:-------------------------------------------------------:	|:--------------:	|:-------------:	|:-------------------------------------------------------------------------------------------------:	|
| df1  	|  South German credit  	|       1000     	|      21     	| [Dataset](https://archive.ics.uci.edu/ml/datasets/South+German+Credit) 	|
| df2  	|        Lung Cancer Data Set       	|       32       	|       56       	|       [Dataset](https://archive.ics.uci.edu/ml/datasets/Container+Crane+Controller+Data+Set)      	|
| df3  	|               Wine Quality Data Set               	|       4898   	|       12       	|              [Dataset](https://archive.ics.uci.edu/ml/datasets/Wine+Quality)              	|

Debido a su disparidad de forma y datos, tomamos estos conjuntos de datos para realizar su respectivo estudio, tomamos tambien data sets no muy amplios, ya que para el procesamiento se el tiempo de espera, al realizar la comparación, no daba el suficiente tiempo por la capacidad de la maquina. 

In [ ]:
# importamos librerias necesarias para el programa

import pandas as pd
import numpy as np
print("Basic libraries Imported Succesfully 🚀")
print("...")

# Libreria para SVM
# importamos el paquete de sklearn necesario para realizar machine learning 
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn import svm
print("Sklearn SVM Imported 🚧")
print("...")

import matplotlib.pyplot as plt
import matplotlib.mlab as mlab

import seaborn
print("Libraries for data visualization 🗠")


Basic libraries Imported Succesfully 🚀
...
Sklearn SVM Imported 🚧
...
Libraries for data visualization 🗠


/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [ ]:
# Importamos google drive para acceder a los conjuntos de datos que estan allí

from google.colab import drive
drive.mount('/content/drive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/drive


In [ ]:
# importamos el dataframe id: df1
df1 = pd.read_csv('/content/drive/My Drive/Colab Notebooks/Datasets/SouthGermanCredit.asc')
#df1 = df1.drop(['Date'], axis=1)
#df1 = df1.drop(['Time'], axis=1)
#df1 = df1.drop(['Unnamed: 15'], axis=1)
#df1 = df1.drop(['Unnamed: 16'], axis=1)
#df1.fillna(0, inplace=True)
#df1.set_index('Unnamed: \0')
df1

laufkont  laufzeit  moral  verw  ...  pers  telef  gastarb  kredit
0           1        18      4     2  ...     2      1        2       1
1           1         9      4     0  ...     1      1        2       1
2           2        12      2     9  ...     2      1        2       1
3           1        12      4     0  ...     1      1        1       1
4           1        12      4     0  ...     2      1        1       1
..        ...       ...    ...   ...  ...   ...    ...      ...     ...
995         1        24      2     3  ...     1      1        2       0
996         1        24      2     0  ...     2      1        2       0
997         4        21      4     0  ...     2      2        2       0
998         2        12      2     3  ...     2      2        2       0
999         1        30      2     2  ...     2      1        2       0

[1000 rows x 21 columns]

In [ ]:
# importamos el dataframe id: df2
df2 = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/lung-cancer/lung-cancer.data')
df2.replace('?',0, inplace=True)
df2.replace('digits',0, inplace=True)
df2

1  0  3  0.1  ?  0.2  2  ...  2.28  2.29  1.12  1.13  1.14  2.30  2.31
0   1  0  3    3  1    0  3  ...     2     2     2     2     1     2     2
1   1  0  3    3  2    0  3  ...     2     2     2     2     2     1     2
2   1  0  2    3  2    1  3  ...     2     2     2     2     2     2     2
3   1  0  3    2  1    1  3  ...     2     2     2     2     1     2     2
4   1  0  3    3  2    0  3  ...     2     2     2     2     2     1     2
5   1  0  3    2  1    0  3  ...     2     1     2     2     2     1     2
6   1  0  2    2  1    0  3  ...     2     2     2     2     1     2     2
7   1  0  3    1  1    0  3  ...     2     2     2     2     1     2     2
8   2  0  2    3  2    0  2  ...     1     3     2     1     1     2     2
9   2  0  2    2  0    0  3  ...     2     2     2     2     2     2     2
10  2  0  2    3  2    0  1  ...     2     2     1     1     2     2     1
11  2  0  2    1  1    0  1  ...     2     2     2     2     1     2     2
12  2  0  2    2  1    1  2  ...     2     2     1     1     1     2     2
13  2  1  3    0  0    1  1  ...     2     2     2     2     1     2     1
14  2  0  3    2  2    1  2  ...     2     2     2     2     2     2     2
15  2  0  3    2  2    0  1  ...     2     2     2     2     1     2     2
16  2  0  2    1  1    0  2  ...     2     2     1     1     1     2     2
17  2  0  2    0  0    0  2  ...     2     2     2     2     2     1     2
18  2  0  1    2  1    0  3  ...     2     2     1     1     2     2     1
19  2  0  2    0  0    1  3  ...     2     1     2     2     1     2     2
20  2  0  3    3  2    0  2  ...     2     2     2     2     2     1     2
21  3  0  2    3  1    1  2  ...     3     1     3     3     2     2     1
22  3  0  2    3  1    1  1  ...     2     2     2     2     2     2     1
23  3  0  3    3  1    0  3  ...     2     3     2     2     2     2     1
24  3  0  2    3  2    0  1  ...     1     3     1     2     2     1     2
25  3  0  2    2  2    0  2  ...     2     2     2     2     1     2     1
26  3  0  2    2  1    0  2  ...     2     3     2     2     2     2     1
27  3  0  3    2  2    0  2  ...     3     1     2     2     2     2     2
28  3  0  2    1  1    0  2  ...     2     2     2     2     2     2     1
29  3  0  2    3  2    1  2  ...     2     2     2     2     1     2     2
30  3  0  2    3  1    0  2  ...     2     2     2     2     2     2     2

[31 rows x 57 columns]

In [ ]:
# importamos el dataframe id: df3
df3 = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/wine-quality/winequality-red.csv',sep=';')
#df3.replace('continuous',0, inplace=True)
df3

fixed acidity  volatile acidity  citric acid  ...  sulphates  alcohol  quality
0               7.4             0.700         0.00  ...       0.56      9.4        5
1               7.8             0.880         0.00  ...       0.68      9.8        5
2               7.8             0.760         0.04  ...       0.65      9.8        5
3              11.2             0.280         0.56  ...       0.58      9.8        6
4               7.4             0.700         0.00  ...       0.56      9.4        5
...             ...               ...          ...  ...        ...      ...      ...
1594            6.2             0.600         0.08  ...       0.58     10.5        5
1595            5.9             0.550         0.10  ...       0.76     11.2        6
1596            6.3             0.510         0.13  ...       0.75     11.0        6
1597            5.9             0.645         0.12  ...       0.71     10.2        5
1598            6.0             0.310         0.47  ...       0.66     11.0        6

[1599 rows x 12 columns]

<hr>

### Estadisticas descriptivas

Realizamos el proceso de estadisticas descriptivas para ver como realizaremos el proceso de machine learning.

In [ ]:
df1.describe()

laufkont     laufzeit  ...      gastarb       kredit
count  1000.000000  1000.000000  ...  1000.000000  1000.000000
mean      2.577000    20.903000  ...     1.963000     0.700000
std       1.257638    12.058814  ...     0.188856     0.458487
min       1.000000     4.000000  ...     1.000000     0.000000
25%       1.000000    12.000000  ...     2.000000     0.000000
50%       2.000000    18.000000  ...     2.000000     1.000000
75%       4.000000    24.000000  ...     2.000000     1.000000
max       4.000000    72.000000  ...     2.000000     1.000000

[8 rows x 21 columns]

In [ ]:
df2.describe()

1          0          3  ...       1.14       2.30       2.31
count  31.000000  31.000000  31.000000  ...  31.000000  31.000000  31.000000
mean    2.064516   0.032258   2.354839  ...   1.580645   1.806452   1.709677
std     0.771815   0.179605   0.550659  ...   0.501610   0.401610   0.461414
min     1.000000   0.000000   1.000000  ...   1.000000   1.000000   1.000000
25%     1.500000   0.000000   2.000000  ...   1.000000   2.000000   1.000000
50%     2.000000   0.000000   2.000000  ...   2.000000   2.000000   2.000000
75%     3.000000   0.000000   3.000000  ...   2.000000   2.000000   2.000000
max     3.000000   1.000000   3.000000  ...   2.000000   2.000000   2.000000

[8 rows x 55 columns]

In [ ]:
df3.describe()

fixed acidity  volatile acidity  ...      alcohol      quality
count    1599.000000       1599.000000  ...  1599.000000  1599.000000
mean        8.319637          0.527821  ...    10.422983     5.636023
std         1.741096          0.179060  ...     1.065668     0.807569
min         4.600000          0.120000  ...     8.400000     3.000000
25%         7.100000          0.390000  ...     9.500000     5.000000
50%         7.900000          0.520000  ...    10.200000     6.000000
75%         9.200000          0.640000  ...    11.100000     6.000000
max        15.900000          1.580000  ...    14.900000     8.000000

[8 rows x 12 columns]

<hr>

## Entrenamiento y comparación

En este paso realizamos el proceso de entrenamiento de maquina y el testeo de la misma para adquirir, una base para poder realizar la comparativa de los resultados. 

## Procedimiento 1 - RBF

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(df1.iloc[:,:20], df1.iloc[:,20:21])

y_train
y_test

clf = SVC(gamma=0.001, C=100)

clf.fit(x_train, y_train)

clf.score(x_test, y_test)
print(clf.score(x_test, y_test))

clf.predict(x_test[:-1])

0.612


/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


array([1, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1,
       0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1,
       1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1,
       1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0,
       0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1,
       1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1, 0, 0,
       1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1,
       1, 0, 0, 1, 0, 1, 0])

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(df2.iloc[:,:55], df2.iloc[:,55:56])
y_train
y_test

clf = SVC(gamma=0.001, C=100)

clf.fit(x_train, y_train)

print(clf.score(x_test, y_test))
clf.score(x_test, y_test)

clf.predict(x_test[:-1])

0.75


/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


array([1, 2, 1, 2, 2, 2, 2])

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(df3.iloc[:,:11], df3.iloc[:,11:12])
y_train
y_test

clf = SVC(gamma=0.001, C=100)

clf.fit(x_train, y_train)

clf.score(x_test, y_test)
print(clf.score(x_test, y_test))

clf.predict(x_test[:-1])

/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


0.565


array([5, 6, 5, 6, 5, 5, 6, 6, 5, 5, 6, 7, 6, 5, 5, 6, 5, 6, 6, 6, 7, 5,
       5, 5, 6, 6, 5, 6, 6, 5, 5, 5, 5, 5, 5, 7, 5, 5, 5, 6, 5, 6, 5, 5,
       5, 5, 5, 5, 5, 6, 6, 6, 6, 6, 5, 6, 5, 6, 5, 6, 5, 5, 6, 5, 5, 5,
       6, 5, 6, 6, 5, 6, 6, 5, 5, 5, 5, 5, 7, 6, 6, 5, 6, 6, 6, 6, 6, 6,
       6, 6, 5, 6, 5, 6, 5, 6, 5, 6, 7, 5, 6, 5, 5, 5, 5, 6, 6, 6, 6, 6,
       7, 5, 5, 6, 5, 5, 6, 7, 6, 5, 6, 6, 6, 7, 5, 6, 6, 5, 5, 6, 5, 6,
       6, 5, 6, 5, 6, 5, 5, 5, 5, 6, 6, 6, 5, 5, 6, 5, 5, 5, 5, 5, 6, 5,
       5, 6, 6, 5, 5, 5, 5, 6, 5, 5, 6, 6, 5, 6, 6, 6, 6, 6, 6, 5, 6, 6,
       5, 5, 5, 6, 6, 6, 7, 5, 6, 5, 6, 6, 6, 5, 5, 7, 5, 5, 7, 6, 7, 6,
       5, 5, 6, 6, 5, 6, 5, 6, 5, 7, 5, 5, 6, 6, 5, 5, 5, 5, 6, 5, 6, 6,
       5, 5, 6, 6, 5, 5, 6, 5, 6, 5, 6, 6, 5, 5, 5, 5, 5, 6, 5, 7, 6, 6,
       6, 6, 6, 5, 5, 6, 5, 6, 6, 5, 6, 6, 5, 5, 6, 5, 5, 6, 7, 5, 5, 5,
       6, 6, 5, 7, 6, 5, 5, 6, 6, 6, 6, 6, 7, 6, 5, 6, 5, 7, 6, 6, 6, 5,
       5, 6, 5, 5, 5, 6, 6, 6, 6, 6, 6, 5, 6, 5, 6,


## Procedimiento 2 - Polinomial

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(df1.iloc[:,:20], df1.iloc[:,20:21])

y_train
y_test

clf = SVC(kernel='poly',C=100)

clf.fit(x_train, y_train)

clf.score(x_test, y_test)
print(clf.score(x_test, y_test))

clf.predict(x_test[:-1])

/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


0.712


array([1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1])

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(df2.iloc[:,:55], df2.iloc[:,55:56])
y_train
y_test

clf = SVC(kernel='poly', C=100)

clf.fit(x_train, y_train)

print(clf.score(x_test, y_test))
clf.score(x_test, y_test)

clf.predict(x_test[:-1])

0.75


/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


array([2, 2, 1, 2, 2, 2, 2])

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(df3.iloc[:,:11], df3.iloc[:,11:12])
y_train
y_test

clf = SVC(kernel='poly', C=100)

clf.fit(x_train, y_train)

clf.score(x_test, y_test)
print(clf.score(x_test, y_test))

clf.predict(x_test[:-1])

/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


0.5475


array([6, 6, 5, 6, 6, 6, 6, 5, 5, 6, 6, 6, 5, 5, 5, 6, 6, 6, 5, 5, 6, 6,
       5, 6, 5, 5, 6, 6, 5, 6, 6, 6, 6, 6, 6, 5, 6, 6, 5, 6, 6, 5, 6, 6,
       6, 6, 6, 6, 6, 6, 6, 6, 5, 6, 6, 5, 5, 6, 5, 5, 6, 5, 6, 5, 6, 6,
       6, 6, 6, 6, 6, 5, 6, 6, 6, 6, 6, 6, 5, 6, 6, 5, 5, 6, 6, 5, 6, 6,
       5, 5, 6, 6, 5, 6, 5, 5, 6, 5, 6, 6, 5, 6, 6, 6, 5, 6, 6, 5, 6, 5,
       6, 5, 6, 5, 5, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 5, 6, 6, 6, 5, 5, 6,
       5, 6, 5, 6, 6, 6, 6, 6, 6, 6, 5, 5, 6, 5, 6, 6, 6, 5, 6, 6, 6, 5,
       5, 5, 6, 6, 6, 6, 6, 6, 5, 5, 5, 5, 6, 6, 5, 6, 6, 6, 5, 6, 6, 6,
       6, 5, 6, 6, 6, 6, 5, 5, 6, 6, 6, 5, 6, 6, 6, 6, 7, 6, 6, 6, 6, 6,
       6, 5, 5, 5, 6, 6, 5, 6, 6, 5, 6, 5, 6, 6, 6, 6, 6, 6, 6, 5, 6, 6,
       6, 5, 6, 6, 6, 5, 6, 6, 5, 5, 6, 6, 5, 5, 6, 6, 6, 6, 6, 6, 6, 6,
       5, 6, 5, 6, 5, 6, 5, 5, 6, 6, 5, 6, 6, 6, 6, 6, 6, 5, 5, 5, 6, 6,
       6, 6, 6, 5, 6, 5, 6, 6, 5, 6, 6, 6, 6, 7, 6, 6, 6, 6, 5, 5, 6, 5,
       5, 6, 6, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 5,

## Procedimiento 3 - Lineal

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(df1.iloc[:,:20], df1.iloc[:,20:21])

y_train
y_test

clf = svm.LinearSVC()

clf.fit(x_train, y_train)

clf.score(x_test, y_test)
print(clf.score(x_test, y_test))

clf.predict(x_test[:-1])

0.716


/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.6/dist-packages/sklearn/svm/_base.py:947: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1])

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(df2.iloc[:,:55], df2.iloc[:,55:56])

y_train
y_test

clf = svm.LinearSVC()

clf.fit(x_train, y_train)

clf.score(x_test, y_test)
print(clf.score(x_test, y_test))

clf.predict(x_test[:-1])

1.0


/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


array([2, 2, 2, 2, 2, 2, 2])

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(df3.iloc[:,:11], df3.iloc[:,11:12])
y_train
y_test

clf = svm.LinearSVC()

clf.fit(x_train, y_train)

clf.score(x_test, y_test)
print(clf.score(x_test, y_test))

clf.predict(x_test[:-1])

/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


0.4625


/usr/local/lib/python3.6/dist-packages/sklearn/svm/_base.py:947: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


array([7, 5, 5, 5, 5, 5, 6, 4, 5, 5, 5, 5, 5, 5, 4, 6, 6, 5, 6, 6, 7, 5,
       5, 6, 6, 6, 6, 6, 5, 5, 6, 5, 5, 5, 5, 6, 7, 5, 7, 7, 5, 5, 6, 6,
       6, 6, 5, 7, 7, 5, 6, 6, 7, 6, 6, 5, 5, 7, 6, 5, 5, 5, 4, 7, 7, 6,
       7, 6, 5, 7, 7, 6, 7, 7, 6, 6, 4, 6, 5, 5, 6, 5, 6, 5, 7, 6, 5, 5,
       7, 7, 6, 5, 7, 6, 5, 5, 7, 5, 7, 6, 5, 5, 6, 6, 6, 5, 6, 5, 7, 7,
       5, 5, 5, 5, 5, 7, 6, 7, 5, 4, 4, 7, 6, 7, 6, 7, 6, 7, 5, 7, 5, 5,
       5, 6, 5, 5, 6, 5, 5, 4, 7, 5, 6, 5, 7, 7, 5, 5, 7, 5, 6, 7, 7, 5,
       5, 5, 5, 5, 6, 5, 5, 6, 7, 6, 5, 5, 6, 4, 5, 5, 5, 6, 6, 7, 6, 5,
       5, 7, 7, 6, 6, 7, 5, 4, 7, 7, 5, 5, 6, 5, 3, 7, 5, 6, 6, 5, 6, 4,
       5, 7, 5, 5, 4, 6, 5, 4, 6, 7, 7, 7, 7, 5, 5, 5, 5, 5, 5, 7, 6, 7,
       5, 5, 6, 5, 5, 6, 5, 6, 6, 7, 6, 5, 7, 5, 7, 5, 6, 6, 6, 7, 6, 6,
       5, 5, 6, 5, 6, 5, 5, 7, 5, 6, 5, 6, 6, 5, 6, 7, 5, 5, 5, 6, 6, 5,
       6, 7, 6, 6, 5, 5, 6, 7, 5, 6, 4, 5, 7, 6, 6, 5, 5, 6, 5, 5, 7, 5,
       6, 7, 5, 5, 5, 6, 5, 4, 5, 7, 5, 6, 7, 5, 7,

## Procedimiento 4 - Sigmoidal

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(df1.iloc[:,:20], df1.iloc[:,20:21])

y_train
y_test

clf = SVC(kernel='sigmoid',C=100)

clf.fit(x_train, y_train)

clf.score(x_test, y_test)
print(clf.score(x_test, y_test))

clf.predict(x_test[:-1])

0.616


/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


array([1, 0, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0,
       1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1,
       1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1,
       1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1, 1,
       1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0,
       0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1,
       0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 1, 1,
       1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1,
       1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1,
       1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1])

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(df2.iloc[:,:55], df2.iloc[:,55:56])
y_train
y_test

clf = SVC(kernel='sigmoid', C=100)

clf.fit(x_train, y_train)

print(clf.score(x_test, y_test))
clf.score(x_test, y_test)

clf.predict(x_test[:-1])

0.75


/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


array([2, 2, 2, 2, 2, 2, 2])

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(df3.iloc[:,:11], df3.iloc[:,11:12])
y_train
y_test

clf = SVC(kernel='sigmoid', C=100)

clf.fit(x_train, y_train)

clf.score(x_test, y_test)
print(clf.score(x_test, y_test))

clf.predict(x_test[:-1])

0.36


/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


array([6, 5, 5, 5, 6, 5, 6, 5, 6, 6, 5, 5, 6, 5, 5, 6, 6, 6, 6, 5, 5, 5,
       3, 5, 5, 5, 6, 6, 5, 5, 6, 5, 6, 6, 5, 6, 6, 3, 6, 6, 6, 6, 5, 6,
       6, 5, 5, 6, 5, 5, 6, 6, 5, 5, 5, 3, 6, 5, 5, 6, 6, 6, 6, 5, 3, 5,
       5, 5, 5, 5, 5, 5, 6, 5, 3, 5, 5, 6, 3, 5, 6, 5, 6, 6, 5, 3, 5, 6,
       3, 5, 6, 5, 6, 5, 5, 5, 6, 6, 6, 6, 5, 6, 6, 6, 5, 5, 6, 6, 5, 5,
       5, 5, 6, 5, 3, 5, 6, 5, 5, 6, 5, 5, 5, 5, 5, 5, 6, 6, 5, 6, 5, 5,
       5, 6, 5, 5, 6, 6, 5, 6, 5, 6, 6, 6, 6, 3, 6, 5, 5, 6, 6, 5, 5, 5,
       6, 6, 6, 5, 6, 5, 6, 5, 5, 6, 5, 6, 5, 6, 5, 6, 6, 5, 3, 6, 6, 6,
       5, 6, 5, 6, 6, 6, 6, 6, 5, 6, 5, 5, 5, 5, 6, 6, 5, 5, 5, 5, 5, 6,
       6, 6, 5, 5, 5, 5, 5, 5, 5, 5, 6, 6, 5, 6, 5, 6, 6, 5, 6, 5, 5, 6,
       5, 6, 6, 5, 6, 5, 6, 6, 5, 5, 6, 5, 5, 5, 5, 5, 3, 5, 6, 3, 6, 5,
       6, 5, 6, 6, 5, 5, 6, 5, 5, 6, 5, 5, 5, 6, 6, 5, 5, 6, 5, 5, 6, 6,
       5, 5, 3, 6, 5, 6, 6, 6, 6, 5, 5, 5, 6, 5, 5, 6, 5, 6, 5, 6, 6, 5,
       5, 5, 5, 5, 3, 5, 5, 6, 5, 5, 5, 6, 5, 5, 5,

<hr>

### Conclusión

Podemos concluir de la practica lo siguiente, para organizar la información la arreglamos de forma tabulada.

| *ID* 	|      **Nombre**     	|  **RBF**  	| **Polinomial** 	| **Lineal** 	| **Sigmoidal** 	|
|------	|:-------------------:	|:---------:	|:--------------:	|:----------:	|:-------------:	|
| df1  	| South German credit 	|   61.2%   	|      71.2%     	|  **71.6%** 	|     61.6%     	|
| df2  	|     Lung cancer     	|    75%    	|       75%      	|  **100%**  	|      75%      	|
| df3  	|     wine quality    	| **56.5%** 	|     54.75%     	|   46.25%   	|      36%      	|

Como podemos ver para cada caso, un proceso puede variar de uno en otro y adquirir mejores resultados, el dato importante o que resalta del resto es en el que despues de realizar el proceso de machine learning, nos dio 100% que para el caso del segundo data frame es el mejor y optimo algoritmo, para su procesamiento es el lineal. Para cada dataFrame puede variar igualmente.

Para poder determinar esto en pocas palabras, podemos realizar un proceso de prueba y error, para ver que tipo de nucleo nos beneficia más.

<hr>

## Opinión pros y contras

### Contras

*   Es difícil predecir el modelo final, ya que puede haber muchos cambios minuciosos. Entonces, recalibrar el modelo cada vez no es una solución.
*   Si hay más características que muestras en los datos, el modelo tendrá un bajo rendimiento.
*   SVM toma mucho tiempo para entrenar con grandes conjuntos de datos.
*   El ajuste de los parámetros SVM no es fácil. Su efecto en el modelo es difícil de ver.
*   Requiere mucha memoria.
*   Difícil de interpretar

### Pros

*    alta precisión
*    buenas garantías teóricas con respecto al sobreajuste
*    sin requisito de distribución
*    selección flexible de núcleos para correlación no lineal
*    Incluso si los datos de entrada no son lineales ni separables, los SVM generan resultados de clasificación precisos debido a su solidez.
*    En la función de decisión, utiliza un subconjunto de puntos de entrenamiento llamados vectores de soporte, por lo tanto, es eficiente en memoria.
*    Es útil para resolver cualquier problema complejo con una función de núcleo adecuada.
*    En la práctica, los modelos SVM son generalizados, con menos riesgo de sobreajuste en SVM.
*    Los SVM funcionan muy bien para la clasificación de texto y para encontrar el mejor separador lineal.
*    Las SVM tienen mejores resultados en producción que las ANN.
*    Pueden manejar eficientemente datos de dimensiones superiores e inseparables linealmente. Son bastante eficientes en memoria.
*    Los problemas complejos se pueden resolver utilizando las funciones del núcleo en la SVM. Esto viene bajo el truco del kernel, que es un gran activo para SVM.
*    Podemos agregar diferentes funciones del núcleo juntas para lograr hiperplanos más complejos.

<hr>

                                                    Fin Taller 8 - KNN MACHINE LEARNING🤖
 
> Juan Felipe Rodríguez Galindo  - **COD. 20181020158**

## Experimentos 🧪 ⚗️

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(df1.iloc[:,:20], df1.iloc[:,20:21])

y_train
y_test

clf = svm.SVR()

clf.fit(x_train, y_train)

clf.score(x_test, y_test)
print(clf.score(x_test, y_test))

clf.predict(x_test[:-1])

-0.1275898930259114


/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


array([0.10938638, 0.40651855, 0.89979682, 0.89953089, 0.89994709,
       0.89939655, 0.90017509, 0.90053236, 0.90003168, 0.89972984,
       0.90039004, 0.67602239, 0.89980726, 0.8998124 , 0.89996687,
       0.86686812, 0.89807274, 0.89977397, 0.89971548, 0.89974222,
       0.89991928, 0.89972417, 0.90045693, 0.89980956, 0.90025313,
       0.8997203 , 0.90032631, 0.8997299 , 0.90570536, 0.89957156,
       0.90000359, 0.8999782 , 0.89975308, 0.89978048, 0.89983164,
       0.90060719, 0.89956835, 0.89981156, 0.89999995, 0.89971292,
       0.64926332, 0.90031509, 0.8999662 , 0.89977844, 0.9000119 ,
       0.89987529, 0.89897465, 0.90010638, 0.89994178, 0.9005763 ,
       0.84535515, 0.89958154, 0.89825016, 0.89979573, 0.89987613,
       0.16816587, 0.89983181, 0.9005845 , 0.9003715 , 0.90387168,
       0.8998475 , 0.90025585, 0.89998178, 0.90012154, 0.8998583 ,
       0.90014139, 0.90037789, 0.90040323, 0.90008846, 0.89963093,
       0.89996322, 0.89962697, 0.90036168, 0.90019099, 0.89971